code/modules/networks/computer3/mainframe2/telesci.dm

```diff
@@ -1234,6 +969,148 @@ TYPEINFO(/obj/machinery/networked/teleconsole)
+			"ztarget" = ztarget,
+			"host_id" = host_id,
+			"readout" = readout,
+			"panel_open" = panel_open,
+			"padNum" = padNum,
+			"max_bookmarks" = max_bookmarks
+		)
+
+		for (var/datum/teleporter_bookmark/b in bookmarks)
+			.["bookmarks"] += list(list(
+				"ref" = ref(b),
+				"name" = b.name,
+				"xyz" = "([b.x]/[b.y]/[b.z]) "
+			))
+		if (!length(bookmarks)) //this is needed to make TGUI clear out the list
+			.["bookmarks"] = list()
+
+	ui_act(action, params)
+		//. = TRUE means the action was handeled
+		. = ..()
+		if (.)
+			return .
+
+		. = TRUE
+
+		playsound(src.loc, 'sound/machines/keypress.ogg', 50, 1, -15)
+
+		switch(action)
+			if ("setX")
+				xtarget = clamp(text2num(params["value"]), 0, 500)
+				coord_update_flag = 1
+			if ("setY")
+				ytarget = clamp(text2num(params["value"]), 0, 500)
+				coord_update_flag = 1
+			if ("setZ")
+				ztarget = clamp(text2num(params["value"]), 0, 14)
+				coord_update_flag = 1
+			if ("send")
+				if (!host_id)
+					boutput(usr, "<span class='alert'>Error: No host connection!</span>")
+					return
+
+				if (coord_update_flag)
+					coord_update_flag = 0
+					message_host("command=teleman&args=-p [padNum] coords x=[xtarget] y=[ytarget] z=[ztarget]")
+
+				message_host("command=teleman&args=-p [padNum] send")
+			if ("receive")
+				if (!host_id)
+					boutput(usr, "<span class='alert'>Error: No host connection!</span>")
+					return
+
+				if (coord_update_flag)
+					coord_update_flag = 0
+					message_host("command=teleman&args=-p [padNum] coords x=[xtarget] y=[ytarget] z=[ztarget]")
+
+				message_host("command=teleman&args=-p [padNum] receive")
+			if ("portal")
+				if (!host_id)
+					boutput(usr, "<span class='alert'>Error: No host connection!</span>")
+					return
+
+				if (coord_update_flag)
+					coord_update_flag = 0
+					message_host("command=teleman&args=-p [padNum] coords x=[xtarget] y=[ytarget] z=[ztarget]")
+
+				message_host("command=teleman&args=-p [padNum] portal toggle")
+			if ("scan")
+				if (!host_id)
+					boutput(usr, "<span class='alert'>Error: No host connection!</span>")
+					return
+
+				if (coord_update_flag)
+					coord_update_flag = 0
+					message_host("command=teleman&args=-p [padNum] coords x=[xtarget] y=[ytarget] z=[ztarget]")
+
+				message_host("command=teleman&args=-p [padNum] scan")
+			if ("restorebookmark")
+				var/datum/teleporter_bookmark/bm = locate(params["value"]) in bookmarks
+				if(!bm) return
+				xtarget = bm.x
+				ytarget = bm.y
+				ztarget = bm.z
+				coord_update_flag = 1
+
+			if ("deletebookmark")
+				var/datum/teleporter_bookmark/bm = locate(params["value"]) in bookmarks
+				if(!bm) return
+				bookmarks.Remove(bm)
+
+			if ("addbookmark")
+				if(length(bookmarks) >= max_bookmarks)
+					boutput(usr, "<span class='alert'>Maximum number of Bookmarks reached.</span>")
+					return
+				var/datum/teleporter_bookmark/bm = new
+				var/title = tgui_input_text(usr, "Enter name:", "Name", "New Bookmark")
```

Comment:
This is probably "fine", but might be neater to have it as a simple text input as part of the UI rather than a popup?

Reply:
Bad two minute MS Paint mockup here:
![image](https://github.com/goonstation/goonstation/assets/62817778/1576687f-6d31-42ea-9b68-a628ca13e84f)